File: leetcode/#94.py
# question --> https://leetcode.com/problems/binary-tree-inorder-traversal/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.stack = list()
        
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans = list()
        while self.stack or root: 
            if root: 
                self.stack.append(root)
                root = root.left
            else:
                root = self.stack.pop()
                ans.append(root.val)
                root = root.right
                
        return ans
        
        # TC: O(N); SC: O(N)
        
# end 
    
    
        
        

File: leetcode/#94.js
// question --> https://leetcode.com/problems/binary-tree-inorder-traversal/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
    if (!root) {
        return [];
    }
    const ans = [];
    const stack = [];
    while (root || stack.length) {
        if (root) {
            stack.push(root); 
            root = root.left; 
        } else {
            root = stack.pop();
            ans.push(root.val); 
            root = root.right;
        } 
    }
    return ans; 
};

// TC: O(N); SC: O(N)

// end 